pkg/submitter/submitter.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2020 The SQLFlow Authors. All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package submitter
+
+import (
+	"fmt"
+	"sort"
+
+	"sqlflow.org/sqlflow/pkg/database"
+	"sqlflow.org/sqlflow/pkg/pipe"
+	"sqlflow.org/sqlflow/pkg/sql/ir"
+
+	pb "sqlflow.org/sqlflow/pkg/proto"
+)
+
+var submitterRegistry = make(map[string]Submitter)
+
+// Submitter submites the Python program by code generator: `pkg/sql/codegen`
+type Submitter interface {
+	ir.Executor
+	setup(*pipe.Writer, *database.DB, string, string, *pb.Session)
+	GetTrainStmtFromModel() bool
+}
+
+// New returns a registered submitter implementation
+func New(name string, w *pipe.Writer, db *database.DB, modelDir, cwd string, session *pb.Session) (Submitter, error) {
+	submitter, ok := submitterRegistry[name]
+	if !ok {
+		return nil, fmt.Errorf("submittter: %s has not been registered", name)
+	}
+	submitter.setup(w, db, modelDir, cwd, session)
+	return submitter, nil
+}
+
+// Register makes a submitter available by the providing name,
+// if the submitter has already registered or registered twice,
+// it panics.
+func Register(name string, submitter Submitter) {
```

Comment:
I am not sure why would we need to build a registry here? If we want to map the name of a submitter type into a SubmitterExecutor, or, simply Executor, implementation, we could use a map.

```go
func NewExecutor(submitter_type string) *Executor {
    m := map[string]func()*Executor {
        "tensorflow": newTensorFlowExecutor,
        "xgboost": newXGBoostExecutor
    }
    if ex, ok := m[submitter_type]; ok {
        return ex
    }
    log.Fatalf("NewExecutor: unknown submitter type %s", submitter_type)
}
```

Reply:
Still, I have a question in my mind -- do we need multiple types of executors? After all, whatever AI APIs a submitter calls, it is a normal Python program. Does that mean we don't need to distinguish the `submitter_type`?

Reply:
> I am not sure why would we need to build a registry here?

All the submitters don't need to edit the `submitter.go` and only add one `init()` function to register itself, such as:

``` golang
# pai_submitter.go
func init() { Register("pai", &paiSubmitter{}) }
```

Reply:
> Still, I have a question in my mind -- do we need multiple types of executors? After all, whatever AI APIs a submitter calls, it is a normal Python program. Does that mean we don't need to distinguish the submitter_type?

For my opinion, we need the different `submitter_type`, `codegen` can generate a Python program, but we need different  submitter to run the AI task, for example, TensorFlow codegen would generate Tensorflow Estimator Program, we can run it by `python train.py` simply, but in the production environment, we also need to  
- submite it as a PAI-Tensorflow task by `pai --name tensorflow1120 -Djobname=sqlflow-train`
- submite it as a PAI-Tensorflow task by Alisa Server: `alisa.createTask(...)`
- submite it as a EDL task by `elasticdl`

So we need different submitter implementation, to execute the program.

Reply:
I see. The model_fn of the Estimator program can come from the model zoo. The Estimator program, or the submitter, is from a codegen. And we need a tool (in order to differentiate from the existing concept submitter, let us use the name runner) to start the Estimator program.

How about we do this -- the submitter program contains the Estimator and its runner.  For example, `pkg/step/codegen/pai.go` generates a Python submitter program which includes an Estimator class whose model_fn is from a model zoo and runs the Estimator class using PAI Python API or the `pai` command-line tool.

Reply:
How about this -- let's have pacakge `pkg/step/codegen`, which contains several implementations of the `CodeGen` interface:

```go
type CodeGen interface {
    Train(*ir.TrainStmt) string
    Predict(*ir.PredictStmt) string
    Explain(*ir.ExplainStmt) string
    Evaluate(*ir.EvaluateStmt) string
}
```

Each implementation, for example, `elasticDLCodeGen`, should have corresponding function `newElasticDLCodeGen`, which can do setup/initialization work.  This is why the `Executor.setup` in https://github.com/sql-machine-learning/sqlflow/pull/1610#discussion_r362692562 is not needed.